File: leetcode/p00098/p00098.py
from typing import Optional, List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def helper(node: Optional[TreeNode]) -> List[int]:
            if not node:
                return []
            return helper(node.left) + [node.val] + helper(node.right)
        
        res = helper(root)
        for i in range(1, len(res)):
            if res[i] <= res[i-1]:
                return False
        return True

File: leetcode/p00098/p00098.go
package p00098

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Time Complexity: O(n), where n is the number of tree nodes
// Space Complexity: O(logk), where k is the max-depth of tree
func isValidBST(root *TreeNode) bool {
	return dfs(root, math.MinInt, math.MaxInt)
}

func dfs(root *TreeNode, lower, upper int) bool {
	if root == nil {
		return true
	}

	if root.Val <= lower || root.Val >= upper {
		return false
	}

	return dfs(root.Left, lower, root.Val) && dfs(root.Right, root.Val, upper)
}

/**
BST 特性, left subtree 的最大值一定會比 root 還小, right subtree 的最小值一定會比最大值還大
所以可以用 dfs 把 lower bound and upper bound 定好之後往下傳遞, 只要不合法就提早結束
*/
